### PR TITLE
A way to implement proper game edit

### PR DESCRIPTION
**Pull request by gmltA**

### CODE REVIEW

File: BsuirFind.Foundation/Services/GameService.cs [hunk 1]
```diff
@@ -60,7 +60,39 @@ public GameService(IBsuirFindUnitOfWork uow, IMapper mapper)
...
+            foreach (var taskToUpdate in tasksToUpdate)
+            {
+                var taskToUpdateDTO = tasksToUpdateDTO.Where(gt => gt.Id == taskToUpdate.Id).SingleOrDefault();
+                if (taskToUpdateDTO == null)
+                {
+                    return ServiceResponse.CreateUnsuccessful<ReadAdminGameDto>(ServiceResponseCode.TaskNotFound);
+                }
+                //todo: WIP. We have to modify other values too.
+                //      Or, maybe, it's better to move this to a function in GameTask class.
+                taskToUpdate.Cost = taskToUpdateDTO.Cost;
```
Comment (gmltA) [inline BsuirFind.Foundation/Services/GameService.cs]: @MikhailSkirukha 
Can we create a function in `GameTask` class which will take another `GameTask` object and update it's own properties according to changes in another object?

File: BsuirFind.Foundation/Services/GameService.cs [hunk 2]
```diff
@@ -52,15 +52,47 @@ public GameService(IBsuirFindUnitOfWork uow, IMapper mapper)
...
+
             var tasksToDelete = game.GameTasks.Where(gt => !gameDto.GameTasks.Select(t => t.Id).Contains(gt.Id)).ToList();
+
+            var tasksToAddDTO = gameDto.GameTasks.Where(gt => gt.Id == 0).ToList();
+            var tasksToAdd = _mapper.Map<IReadOnlyCollection<EditGameTaskDto>, IReadOnlyCollection<GameTask>>(tasksToAddDTO);
+
+            foreach (var taskToAdd in tasksToAdd)
+            {
+                // todo: we query DB for each task. That's not right. Maybe we can batch this process within one reuqest?
+                var task = await _uow.Tasks.GetByIdAsync(taskToAdd.TaskId);
```
Comment (zaq007) [inline BsuirFind.Foundation/Services/GameService.cs]: To evade such many db requests here you can just replace
`
var task = await _uow.Tasks.GetByIdAsync(taskToAdd.TaskId);
if (task == null)
{
    return ServiceResponse.CreateUnsuccessful<ReadAdminGameDto>(ServiceResponseCode.TaskNotFound);
}
taskToAdd.Task = task;
`
with
`
taskToAdd.TaskId = taskToAdd.TaskId;
`

File: BsuirFind.Foundation/Services/GameService.cs (hunk 1 above)
Comment (MikhailSkirukha) [reply BsuirFind.Foundation/Services/GameService.cs]: @gmltA 
There must be no business logic in domain models.

File: BsuirFind.Foundation/Services/GameService.cs (hunk 2 above)
Comment (zaq007) [reply BsuirFind.Foundation/Services/GameService.cs]: Ooops..
You can remove this at all then :)

Comment (zaq007) [reply BsuirFind.Foundation/Services/GameService.cs]: @gmltA 
Sure.
ID is exactly what you have in DB table so you can assign it explicitly

Comment (gmltA) [reply BsuirFind.Foundation/Services/GameService.cs]: @zaq007 
Nice. I thought that EF won't allow me this.

File: BsuirFind.Repositories/Repositories/Repository.cs [hunk 3]
```diff
@@ -76,6 +83,17 @@ public virtual void Update(TEntity entity)
...
+        public virtual void UpdateRange(IReadOnlyCollection<TEntity> entities)
+        {
+            LoggerContext.Logger.Info($"Update range of entities in database with type {typeof(TEntity).Name}");
+            foreach (var entity in entities) {
+                if (!_dbSet.Local.Contains(entity))
+                {
+                    Context.Entry(entity).State = EntityState.Modified;
+                }
+            }
+        }
```
Comment (MikhailSkirukha) [inline BsuirFind.Repositories/Repositories/Repository.cs]: Unnecessary method. Remove

File: BsuirFind.Repositories.Interfaces/IRepository.cs [hunk 4]
```diff
@@ -32,6 +37,11 @@ public interface IRepository<TEntity> where TEntity : class
         /// </summary>
         void Update(TEntity entity);
 
+        /// <summary>
+        ///     Updates a range of existing entities.
+        /// </summary>
+        void UpdateRange(IReadOnlyCollection<TEntity> entities);
+
```
Comment (MikhailSkirukha) [inline BsuirFind.Repositories.Interfaces/IRepository.cs]: Must be removed

**Pull request merged**